core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/SubmissionCostService.scala

```diff
@@ -17,34 +19,86 @@ object SubmissionCostService {
 
 class SubmissionCostService(tableName: String, serviceProject: String, bigQueryDAO: GoogleBigQueryDAO)(implicit val executionContext: ExecutionContext) extends LazyLogging {
 
+  val stringParamType = new QueryParameterType().setType("STRING")
 
-  def getWorkflowCosts(workflowIds: Seq[String],
-                       workspaceNamespace: String): Future[Map[String, Float]] = {
+  def getSubmissionCosts(submissionId: String, workflowIds: Seq[String], workspaceNamespace: String, submissionDate: Option[DateTime]): Future[Map[String, Float]] = {
+    if( workflowIds.isEmpty ) {
+      Future.successful(Map.empty[String, Float])
+    } else {
+      for {
+        //try looking up the workflows via the submission ID.
+        //this makes for a smaller query string (though no faster).
+        submissionCosts <- executeSubmissionCostsQuery(submissionId, workspaceNamespace, submissionDate)
+        //if that doesn't return anything, fall back to
+        fallbackCosts <- if (submissionCosts.size() == 0)
+          executeWorkflowCostsQuery(workflowIds, workspaceNamespace, submissionDate)
+        else
+          Future.successful(submissionCosts)
+      } yield {
+        extractCostResults(fallbackCosts)
+      }
+    }
+  }
 
-    extractWorkflowCostResults(executeWorkflowCostsQuery(workflowIds, workspaceNamespace))
+
+  def getWorkflowCost(workflowId: String,
+                      workspaceNamespace: String,
+                      submissionDate: Option[DateTime]): Future[Map[String, Float]] = {
+    executeWorkflowCostsQuery(Seq(workflowId), workspaceNamespace, submissionDate) map extractCostResults
   }
 
   /*
    * Manipulates and massages a BigQuery result.
    */
-  def extractWorkflowCostResults(rowsFuture: Future[util.List[TableRow]]): Future[Map[String, Float]] = {
-
-    rowsFuture map { rowsOrNull =>
-      Option(rowsOrNull) match {
-        case Some(rows) => rows.asScala.map { row =>
-          // workflow ID is contained in the 2nd cell, cost is contained in the 3rd cell
-          row.getF.get(1).getV.toString -> row.getF.get(2).getV.toString.toFloat
-        }.toMap
-        case None => Map.empty[String, Float]
-      }
+  def extractCostResults(rows: util.List[TableRow]): Map[String, Float] = {
```

Comment:
I didn't like how this took a future and returned a future when it didn't need to. I moved the map call up to the caller.